cpp/src/glm/qn/glm_base_mg.cuh

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2020-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/core/comms.hpp>
+#include <raft/util/cudart_utils.hpp>
+#include <raft/core/handle.hpp>
+#include <raft/linalg/multiply.cuh>
+
+#include "glm/qn/glm_base.cuh"
+#include "glm/qn/glm_logistic.cuh"
+#include "glm/qn/qn_util.cuh"
+#include "glm/qn/qn_solvers.cuh"
+#include "glm/qn/glm_regularizer.cuh"
+#include <vector>
+#include <iostream>
+
+namespace ML {
+namespace GLM {
+namespace opg {
+template<typename T>
+// multi-gpu version of linearBwd
+inline void linearBwdMG(const raft::handle_t& handle,
+                      SimpleDenseMat<T>& G,
+                      const SimpleMat<T>& X,
+                      const SimpleDenseMat<T>& dZ,
+                      bool setZero,
+                      const int64_t n_samples, 
+                      const int n_ranks)
+{
+  cudaStream_t stream = handle.get_stream();
+  // Backward pass:
+  // - compute G <- dZ * X.T
+  // - for bias: Gb = mean(dZ, 1)
+
+  const bool has_bias = X.n != G.n;
+  const int D         = X.n;
+  const T beta        = setZero ? T(0) : T(1);
+
+  if (has_bias) {
+
+    SimpleVec<T> Gbias;
+    SimpleDenseMat<T> Gweights;
+
+    col_ref(G, Gbias, D);
+
+    col_slice(G, Gweights, 0, D);
+
+    // TODO can this be fused somehow?
+    const SimpleDenseMat<T>* X_simple_p = (const SimpleDenseMat<T>*)(&X);
+    Gweights.assign_gemm(handle, 1.0 / n_samples, dZ, false, X, false, beta / n_ranks, stream);
+
+    raft::stats::mean(Gbias.data, dZ.data, dZ.m, dZ.n, false, true, stream);
+    T bias_factor = 1.0 * dZ.n / n_samples;
+    raft::linalg::multiplyScalar(Gbias.data, Gbias.data, bias_factor, dZ.m, stream);
+
+  } else {
+    CUML_LOG_DEBUG("has bias not enabled");
+    G.assign_gemm(handle, 1.0 / n_samples, dZ, false, X, false, beta / n_ranks, stream);
+  }
+}
+
+template <typename T, class GLMObjective>
+struct GLMWithDataMG : ML::GLM::detail::GLMWithData<T, GLMObjective> {
```

Comment:
Could we have a doxygen here just to help future developers?

Reply:
added, please check!